File: solutions/yoon-wootak/2026-08-04/medium/7465.java
import java.io.*;
import java.util.*;

class Solution {
	static int[] parent;
	
	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("input.txt"));
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		
		int T = Integer.parseInt(br.readLine());
		for (int test_case=1; test_case<=T; test_case++) {
			sb.append("#").append(test_case).append(" ");
			
			StringTokenizer st = new StringTokenizer(br.readLine());
			int N = Integer.parseInt(st.nextToken());
			int M = Integer.parseInt(st.nextToken());
			
			parent = new int[N + 1];
			
			for (int i=1; i<=N; i++) {
				parent[i] = i;
			}
			
			for (int i=0; i<M; i++) {
				st = new StringTokenizer(br.readLine());
				int a = Integer.parseInt(st.nextToken());
				int b = Integer.parseInt(st.nextToken());
				
				union(a, b);
			}
			
			Set<Integer> hs = new HashSet<>();
			for (int i=1; i<=N; i++) {
				hs.add(find(i));
			}	
		
			sb.append(hs.size()).append("\n");
		}
		
		System.out.println(sb);
	}
	
	//부모를 찾는 함수
	static int find(int x) {
		if (parent[x] == x) {
			return x;
		}
		return parent[x] = find(parent[x]);
	}
	
	//부모-자식 관계 추가 함수
	static void union(int a, int b) {
		int rootA = find(a);
		int rootB = find(b);
		
		if (rootA != rootB) {
			parent[rootB] = rootA;
		}
	}
}

File: solutions/SURI913/2026-07-31/medium/7465.cpp
#include <iostream>
using namespace std;
#define MAX 101

int parent[MAX];
int cnt;
//유니온 파인드
void SetParent(int n) {
	for (int i = 0; i <= n; i++) {
		parent[i] = i;
	}
}

int Find(int x) {
	if (parent[x] == x) return x;
	return parent[x] = Find(parent[x]);
}

void Union(int a, int b) {
	int rootA = Find(a);
	int rootB = Find(b);

	//이미 최상단 부모라
	if (rootA != rootB) {
		//A -B
		parent[rootB] = rootA; //parent[rootA]아님 rootA 들어가야지
		cnt--; //그룹 수 세기 위한 cnt 부모 정해질때마다 줄임
	}
}

int main() {
	int T;
	cin >> T;
	for (int test_case = 1; test_case <= T; test_case++) {
		int N, M;
		cin >> N >> M;
		SetParent(N);
		cnt = N;
		int a, b;
		for (int i = 0; i < M; i++) {
			cin >> a >> b;
			Union(a, b);
		}

		cout << '#' << test_case << ' ' << cnt << endl;
	}
}